indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/SeekableStreamIndexTaskRunner.java

```diff
@@ -16,55 +16,1991 @@
+                    snapshot
+                )
+            );
+          }
+
+          @Override
+          public void run()
+          {
+            // Do nothing.
+          }
+        };
+      };
+
+      // restart publishing of sequences (if any)
+      maybePersistAndPublishSequences(committerSupplier);
+
+      Set<StreamPartition<PartitionType>> assignment = assignPartitions(recordSupplier);
+      possiblyResetDataSourceMetadata(recordSupplier, assignment);
+      seekToStartingSequence(recordSupplier, assignment);
+
+      ingestionState = IngestionState.BUILD_SEGMENTS;
+
+      // Main loop.
+      // Could eventually support leader/follower mode (for keeping replicas more in sync)
+      boolean stillReading = !assignment.isEmpty();
+      status = Status.READING;
+      Throwable caughtExceptionInner = null;
+
+      initialOffsetsSnapshot.putAll(currOffsets);
+      exclusiveStartingPartitions.addAll(ioConfig.getExclusiveStartSequenceNumberPartitions());
+
+      try {
+        while (stillReading) {
+          if (possiblyPause()) {
+            // The partition assignments may have changed while paused by a call to setEndOffsets() so reassign
+            // partitions upon resuming. This is safe even if the end sequences have not been modified.
+            assignment = assignPartitions(recordSupplier);
+            possiblyResetDataSourceMetadata(recordSupplier, assignment);
+            seekToStartingSequence(recordSupplier, assignment);
+
+            if (assignment.isEmpty()) {
+              log.info("All partitions have been fully read");
+              publishOnStop.set(true);
+              stopRequested.set(true);
+            }
+          }
+
+          // if stop is requested or task's end sequence is set by call to setEndOffsets method with finish set to true
+          if (stopRequested.get() || sequences.get(sequences.size() - 1).isCheckpointed()) {
+            status = Status.PUBLISHING;
+          }
+
+          if (stopRequested.get()) {
+            break;
+          }
+
+          if (backgroundThreadException != null) {
+            throw new RuntimeException(backgroundThreadException);
+          }
+
+          checkPublishAndHandoffFailure();
+
+          maybePersistAndPublishSequences(committerSupplier);
+
+
+          // calling getReocrd() ensures that excpetions specific to kafka/kinesis like OffsetOutOfRangeException
+          // are handled in the subclasses
+          List<OrderedPartitionableRecord<PartitionType, SequenceType>> records = getRecords(recordSupplier, toolbox);
+
+          stillReading = !assignment.isEmpty();
+
+          SequenceMetadata sequenceToCheckpoint = null;
+          for (OrderedPartitionableRecord<PartitionType, SequenceType> record : records) {
+
+            // for Kafka, the end offsets are exclusive, so skip it
+            if (getRunnerType() == Type.KAFKA && createSequenceNumber(record.getSequenceNumber()).equals(
```

Comment:
Can you explain what the difference between the Kafka and Kinesis implementation is here? Why does Kafka need to skip the end offsets?

Reply:
the behavior for the original kafka implementation is to read from startingOffsets (inclusive) to endOffsets (exclusive), whereas the for kinesis the behavior is to read from start (inclusive) to end (inclusive) due to limitations in getting the next sequence number. This check is added here to ensure similar behavior with the old kafka task implementation. If you look at `KafkaSequenceMetadata`, the behavior of canHandle() is copied from the old kafka task implementation.
```java
    protected boolean canHandle(OrderedPartitionableRecord<Integer, Long> record)
    {
      lock.lock();
      try {
        final OrderedSequenceNumber<Long> partitionEndOffset = createSequenceNumber(endOffsets.get(record.getPartitionId()));
        final OrderedSequenceNumber<Long> partitionStartOffset = createSequenceNumber(startOffsets.get(record.getPartitionId()));
        final OrderedSequenceNumber<Long> recordOffset = createSequenceNumber(record.getSequenceNumber());
        return isOpen()
               && recordOffset != null
               && partitionEndOffset != null
               && partitionStartOffset != null
               && recordOffset.compareTo(partitionStartOffset) >= 0
               && recordOffset.compareTo(partitionEndOffset) < 0;
      }
      finally {
        lock.unlock();
      }
    }
```